vaadin-grid-flow/src/main/resources/META-INF/resources/frontend/gridConnector.js

```diff
@@ -1,1017 +1,1023 @@
         }
 
-        let parentCache = grid.$connector.getCacheByKey(parentUniqueKey);
-        let itemCache = (parentCache && parentCache.itemkeyCaches) ? parentCache.itemkeyCaches[parentUniqueKey] : undefined;
-        if(cache[parentUniqueKey] && cache[parentUniqueKey][page] && itemCache) {
-          // workaround: sometimes grid-element gives page index that overflows
-          page = Math.min(page, Math.floor(cache[parentUniqueKey].size / grid.pageSize));
+        let firstPage = Math.max(0,  firstNeededPage);
+        let lastPage = (parentKey !== root) ? lastNeededPage: Math.min(lastNeededPage, Math.floor(grid.size / grid.pageSize));
+        let lastRequestedRange = lastRequestedRanges[parentKey];
+        if(!lastRequestedRange) {
+          lastRequestedRange = [-1, -1];
+        }
+        if (lastRequestedRange[0] != firstPage || lastRequestedRange[1] != lastPage) {
+          lastRequestedRange = [firstPage, lastPage];
+          lastRequestedRanges[parentKey] = lastRequestedRange;
+          let count = lastPage - firstPage + 1;
+          fetch(firstPage * grid.pageSize, count * grid.pageSize);
+        }
+      })
 
-          callback(cache[parentUniqueKey][page], cache[parentUniqueKey].size);
-        } else {
-          treePageCallbacks[parentUniqueKey][page] = callback;
+      grid.dataProvider = tryCatchWrapper(function(params, callback) {
+        if (params.pageSize != grid.pageSize) {
+          throw 'Invalid pageSize';
         }
-        grid.$connector.fetchPage((firstIndex, size) =>
-            grid.$connector.beforeParentRequest(firstIndex, size, params.parentItem.key),
-            page, parentUniqueKey);
 
-      } else {
-        // workaround: sometimes grid-element gives page index that overflows
-        page = Math.min(page, Math.floor(grid.size / grid.pageSize));
+        let page = params.page;
+
+        if(params.parentItem) {
+          let parentUniqueKey = grid.getItemId(params.parentItem);
+          if(!treePageCallbacks[parentUniqueKey]) {
+            treePageCallbacks[parentUniqueKey] = {};
+          }
+
+          let parentCache = grid.$connector.getCacheByKey(parentUniqueKey);
+          let itemCache = (parentCache && parentCache.itemkeyCaches) ? parentCache.itemkeyCaches[parentUniqueKey] : undefined;
+          if(cache[parentUniqueKey] && cache[parentUniqueKey][page] && itemCache) {
+            // workaround: sometimes grid-element gives page index that overflows
+            page = Math.min(page, Math.floor(cache[parentUniqueKey].size / grid.pageSize));
+
+            callback(cache[parentUniqueKey][page], cache[parentUniqueKey].size);
+          } else {
+            treePageCallbacks[parentUniqueKey][page] = callback;
+          }
+          grid.$connector.fetchPage((firstIndex, size) =>
+              grid.$connector.beforeParentRequest(firstIndex, size, params.parentItem.key),
+              page, parentUniqueKey);
```

Comment:
Indentation seems a bit weird but this would also be more clear imo if all the parameters were on their own lines.  
I'd change this to:
```js
          grid.$connector.fetchPage(
            (firstIndex, size) => grid.$connector.beforeParentRequest(firstIndex, size, params.parentItem.key),
            page,
            parentUniqueKey
          );
```

Reply:
Updated.

Reply:
It's better, but there's still double indentation (4 spaces instead of 2) on the inner lines here (the parameters).

Reply:
Updated.